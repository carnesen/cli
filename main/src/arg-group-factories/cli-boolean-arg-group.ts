import { ICliArgGroup } from '../cli-arg-group';
import { CliUsageError } from '../cli-usage-error';

/**
 * Options for [[`CliBooleanArgGroup`]]
 */
export interface ICliBooleanArgGroupOptions {
	/** See [[`ICliArgGroup.description`]] */
	description?: string;
	/** See [[`ICliArgGroup.hidden`]] */
	hidden?: boolean;
}

/**
 * A factory for boolean-valued [[`ICliArgGroup`]]'s
 *
 * @example
 * ```plaintext
 * $ cli           // named flag "foo" parses `false`
 * $ cli --foo     // named flag "foo" parses `true`
 * $ cli --foo bar // usage error
 * ```
 *
 * @throws [[`CliUsageError`]]
 */
export function CliBooleanArgGroup(
	options: ICliBooleanArgGroupOptions = {},
): ICliArgGroup<boolean, false> {
	const { description, hidden = false } = options;
	const argGroup: ICliArgGroup<boolean, false> = {
		placeholder: '',
		required: false,
		hidden,
		parse(args) {
			if (!args) {
				return false;
			}
			if (args.length > 0) {
				throw new CliUsageError(`Unexpected argument "${args[0]}"`);
			}
			return true;
		},
		description,
	};
	return argGroup;
}
