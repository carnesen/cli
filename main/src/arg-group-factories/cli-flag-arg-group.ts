import { ICliArgGroup } from '../cli-arg-group';
import { CliUsageError } from '../cli-usage-error';
import { TCliDescription } from '../cli-description';

/**
 * Options for [[`CliFlagArgGroup`]]
 */
export interface ICliFlagArgGroupOptions {
	/** See [[`ICliArgGroup.description`]] */
	description?: TCliDescription;
	/** See [[`ICliArgGroup.hidden`]] */
	hidden?: boolean;
}

/**
 * A factory for boolean-valued [[`ICliArgGroup`]]'s
 *
 * @example
 * ```plaintext
 * $ cli           // named flag "foo" parses `false`
 * $ cli --foo     // named flag "foo" parses `true`
 * $ cli --foo bar // usage error
 * ```
 *
 * @throws [[`CliUsageError`]]
 */
export function CliFlagArgGroup(
	options: ICliFlagArgGroupOptions = {},
): ICliArgGroup<boolean, false> {
	const { description, hidden = false } = options;
	const argGroup: ICliArgGroup<boolean, false> = {
		placeholder: '',
		required: false,
		hidden,
		parse(args) {
			if (!args) {
				return false;
			}
			if (args.length > 0) {
				throw new CliUsageError(`Unexpected argument "${args[0]}"`);
			}
			return true;
		},
		description,
	};
	return argGroup;
}
