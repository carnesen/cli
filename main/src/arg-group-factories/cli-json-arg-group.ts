import { ICliArgGroup } from '../cli-arg-group';
import { CliUsageError } from '../cli-usage-error';

/** Options for [[`CliJsonArgGroup`]] */
export interface ICliJsonArgGroupOptions {
	/** [[`ICliArgGroup.description`]] */
	description?: string;

	/** [[`ICliArgGroup.required`]] */
	required?: boolean;

	/** [[`ICliArgGroup.placeholder`]]. Defaults to "\<json\>" */
	placeholder?: string;

	/** [[`ICliArgGroup.hidden`]] */
	hidden?: boolean;
}

/**
 * A factory for [[`ICliArgGroup`]]s that `JSON.parse`
 *
 * @example
 * Suppose our CLI's positionalArgGroup is a CliJsonArgGroup here's how that
 * behaves:
 * ```plaintext
 * $ cli '{"foo":true}' // named value "json" receives object `{ foo: true }`
 * $ cli                // usage error
 * $ cli '""' '""'      // usage error
 * $ cli foo            // error parsing JSON
 * ```
 *
 * @throws [[`CliUsageError`]]
 */
export function CliJsonArgGroup(
	options: ICliJsonArgGroupOptions = {},
): ICliArgGroup<any> {
	const {
		placeholder = '<json>',
		required = false,
		description,
		hidden = false,
	} = options;
	const argGroup: ICliArgGroup<any> = {
		required,
		placeholder,
		hidden,
		parse(args) {
			if (!args) {
				return undefined;
			}
			if (args.length !== 1) {
				throw new CliUsageError(`Expected a single ${placeholder} string`);
			}
			try {
				const parsed = JSON.parse(args[0]);
				return parsed;
			} catch (exception) {
				throw new CliUsageError(exception.message);
			}
		},
		description,
	};
	return argGroup;
}
