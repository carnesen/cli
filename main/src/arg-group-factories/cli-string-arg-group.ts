import { ICliArgGroup } from '../cli-arg-group';
import { CliUsageError } from '../cli-usage-error';

/** Options for [[`CliStringArgGroup`]] */
export interface ICliStringArgGroupOptions {
	/** See [[`ICliArgGroup.required`]] */
	required?: boolean;

	/** See [[`ICliArgGroup.description`]] */
	description?: string;

	/** See [[`ICliArgGroup.hidden`]] */
	hidden?: boolean;

	/** See [[`ICliArgGroup.placeholder`]] defaulting to "\<str\>" */
	placeholder?: string;
}

/** A factory for required `string`-valued [[`ICliArgGroup`]]s */
function CliStringArgGroup(
	options: ICliStringArgGroupOptions & { required: true },
): ICliArgGroup<string, true>;

/** A factory for optional `string | undefined`-valued [[`ICliArgGroup`]]s */
function CliStringArgGroup(
	options?: ICliStringArgGroupOptions,
): ICliArgGroup<string | undefined, false>;

// Implementation
function CliStringArgGroup(
	options: ICliStringArgGroupOptions = {},
): ICliArgGroup<string | undefined> {
	const {
		required = false,
		description,
		placeholder = '<str>',
		hidden = false,
	} = options;
	const argGroup: ICliArgGroup<string | undefined> = {
		hidden,
		placeholder,
		required,
		parse(args) {
			if (!args) {
				return undefined;
			}

			if (args.length > 1) {
				throw new CliUsageError(`Expected just one ${placeholder}`);
			}

			if (args.length === 0) {
				throw new CliUsageError(`Expected a ${placeholder}`);
			}

			return args[0];
		},
		description,
	};
	return argGroup;
}

export { CliStringArgGroup };
