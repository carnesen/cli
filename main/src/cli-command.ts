import { CLI_COMMAND } from './constants';
import {
  AnyArgParser,
  AnyNamedArgParsers,
  ValueFromArgParser,
  NamedValues,
  ICliArgParser,
} from './cli-arg-parser';

export type CliCommandOptions<
  TPositionalArgParser extends AnyArgParser,
  TNamedArgParsers extends AnyNamedArgParsers,
  TEscapedArgParser extends AnyArgParser
> = {
  /**
   * Name used in command-line usage
   */
  name: string;
  /**
   * The function that defines your command logic.
   *
   * @param positionalValue
   * @param namedValues
   * @param escapedValue
   * @returns A value (synchronously) or a `Promise` that resolves to a value
   * @remarks
   * By default, {@linkcode runCliAndExit} `console.log`s the value returned by `action`.
   * The type of the `args` argument received by `action` is derived by the `args`
   * property of the command. Similarly, the `options` argument type is derived from
   * `command.options`.
   *
   */
  action: (
    positionalValue: ValueFromArgParser<TPositionalArgParser>,
    namedValues: NamedValues<TNamedArgParsers>,
    escapedValue: ValueFromArgParser<TEscapedArgParser>,
  ) => any;
  positionalArgParser?: TPositionalArgParser;
  namedArgParsers?: TNamedArgParsers;
  escapedArgParser?: TEscapedArgParser;
  /**
   * A string that will be included in `Usage:` if present.
   */
  description?: string;
  /**
   * If true, don't show this command in usage docs.
   */
  hidden?: boolean;
};

/**
 * Interface describing a "command" command
 * @typeParam TPositionalArgParser Type of the "positional" arguments parser
 * @typeParam TNamedArgParsers Type of the ""
 */
export interface ICliCommand<
  TPositionalArgParser extends AnyArgParser,
  TNamedArgParsers extends AnyNamedArgParsers,
  TEscapedArgParser extends AnyArgParser
> extends CliCommandOptions<TPositionalArgParser, TNamedArgParsers, TEscapedArgParser> {
  commandType: typeof CLI_COMMAND;
}

/**
 *
 * A factory function for creating CLI command commands
 * @param options
 * @returns The newly-created `command`.
 */
export function CliCommand<
  TPositional extends AnyArgParser = ICliArgParser<undefined, false>,
  TNamed extends AnyNamedArgParsers = any,
  TEscaped extends AnyArgParser = ICliArgParser<undefined, false>
>(
  options: CliCommandOptions<TPositional, TNamed, TEscaped>,
): ICliCommand<TPositional, TNamed, TEscaped> {
  return {
    ...options,
    commandType: CLI_COMMAND,
  };
}
