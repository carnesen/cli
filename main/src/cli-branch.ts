import { ICliCommand } from './cli-command';

/** "kind" of an {@linkcode ICliBranch} */
export const CLI_BRANCH = 'CLI_BRANCH';

/** Options for {@linkcode CliBranch} */
export interface ICliBranchOptions {
  /** Name of this command branch, typically a category like "user" or "cloud" */
  name: string;

  /** A short description of this branch for command-line usage */
  description?: string;

  /** If `true`, this branch will not appear in command-line usage */
  hidden?: boolean;

  /** Branches and/or commands underneath this branch */
  children: (ICliBranch | ICliCommand<any, any, any>)[];
}

/**
 * A branch in a command tree
 */
export interface ICliBranch extends ICliBranchOptions {
  /** Used internally for discriminating between {@linkcode ICliBranch}'s and
   * {@linkcode ICliCommand}'s */
  kind: typeof CLI_BRANCH;
}

/**
 * A factory for {@linkcode ICliBranch} objects
 * @param options
 * @returns An {@linkcode ICliBranch} object
 * @example
 * ```typescript
 * const cloudBranch = CliBranch()
 * ```
 */
export function CliBranch(options: ICliBranchOptions): ICliBranch {
  return {
    ...options,
    kind: CLI_BRANCH,
  };
}
