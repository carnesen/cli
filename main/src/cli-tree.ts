import { ICliBranch } from './cli-branch';
import { ICliCommand } from './cli-command';
import { ICliArgGroup } from './cli-arg-group';

/**
 * The root of a command tree
 */
export type TCliRoot =
	| ICliBranch
	| ICliCommand<ICliArgGroup, any, ICliArgGroup>;

/**
 * Data structure representing a node in a command tree
 * */
export interface ICliTree {
	current: TCliRoot;
	parents: ICliBranch[];
}

/**
 * Data structure representing a leaf node in a command tree
 * */
export interface ICliLeaf {
	current: ICliCommand;
	parents: ICliBranch[];
}
