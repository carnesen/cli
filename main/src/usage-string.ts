import redent = require('redent');

import { Node } from './cli-node';
import { CLI_BRANCH } from './cli-branch';
import { TextList } from './text-list';
import { regularizeText, wrapInSquareBrackets } from './util';
import { getPathAndDescriptionOfLeaves } from './get-path-and-description-of-commands';
import { AnyParser } from './cli-arg-parser';

const INDENT_SIZE = 3;

export function UsageString({ current, parents }: Node): string {
  const commandPathString = [...parents, current].map(({ name }) => name).join(' ');
  let firstParagraph = `Usage: ${commandPathString}`;
  const otherParagraphs: string[] = [];

  function appendValuedParserUsage(parser?: AnyParser, prefix?: string) {
    if (parser && !parser.hidden) {
      const { placeholder, description, required } = parser;
      if (prefix) {
        firstParagraph += ` ${prefix}`;
      }
      firstParagraph += ` ${required ? placeholder : wrapInSquareBrackets(placeholder)}`;
      if (description) {
        otherParagraphs.push(`${placeholder}:`);
        otherParagraphs.push(redent(regularizeText(description), INDENT_SIZE));
      }
    }
  }

  if (current.kind === CLI_BRANCH) {
    // BRANCH
    firstParagraph += ' <subcommand> ...';
    otherParagraphs.push('Subcommands:');
    const nameAndDescriptionOfLeaves = getPathAndDescriptionOfLeaves(current, []);
    const items: Parameters<typeof TextList> = nameAndDescriptionOfLeaves.map(
      ({ path, description }) => ({
        name: path.join(' '),
        text: description,
      }),
    );
    const childrenParagraph = redent(TextList(...items), INDENT_SIZE);
    otherParagraphs.push(childrenParagraph);
  } else {
    // LEAF
    const {
      positionalParser: positionalValuedParser,
      namedParsers: namedValuedParsers,
      escapedParser: escapedValuedParser,
    } = current;

    appendValuedParserUsage(positionalValuedParser);

    if (namedValuedParsers) {
      const entries = Object.entries(namedValuedParsers).filter(
        ([_, parser]) => !parser.hidden,
      );
      if (entries.length > 0) {
        const optionsNotRequired = entries.every(
          ([_, namedValuedParser]) => !namedValuedParser.required,
        );
        firstParagraph += optionsNotRequired ? ' [<options>]' : ' <options>';
        otherParagraphs.push('Options:');
        const items: Parameters<typeof TextList> = entries.map(([parserName, parser]) => {
          let name = `--${parserName}`;
          if (parser.placeholder) {
            name += ` ${parser.placeholder}`;
          }
          if (!parser.required) {
            name = wrapInSquareBrackets(name);
          }
          const text = parser.description;
          return { name, text };
        });
        const optionsParagraph = redent(TextList(...items), INDENT_SIZE);
        otherParagraphs.push(optionsParagraph);
      }
    }

    appendValuedParserUsage(escapedValuedParser, '--');
  }

  const paragraphs = [firstParagraph];

  const regularizedDescription = regularizeText(current.description);
  if (regularizedDescription) {
    paragraphs.push(redent(regularizedDescription, INDENT_SIZE));
  }

  paragraphs.push(...otherParagraphs);

  return paragraphs.join('\n\n');
}
