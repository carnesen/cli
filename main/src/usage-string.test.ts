import { UsageString } from './usage-string';
import { CliBranch } from './cli-branch';
import { CliStringArgGroup } from './arg-group-factories/cli-string-arg-group';
import { CliCommand } from './cli-command';
import { ICliNode } from './cli-tree';

const messageArgGroup = CliStringArgGroup({
	description: 'A string message please',
});
const positionalArgGroup = CliStringArgGroup({
	description: 'A word',
	placeholder: '<word>',
});
const escapedArgGroup = CliStringArgGroup({
	description: 'Another word',
	required: true,
});

const current = CliCommand({
	name: 'echo',
	positionalArgGroup,
	namedArgGroups: {
		message: messageArgGroup,
	},
	escapedArgGroup,
	action(foo) {
		return foo;
	},
});

const branch = CliBranch({
	name: 'cli',
	description: 'This is a CLI',
	children: [current],
});

describe(UsageString.name, () => {
	it('Creates a usage string for a branch', () => {
		const usageString = UsageString({ current: branch, parents: [] });
		expect(usageString).toMatchSnapshot();
	});

	it('Creates a usage string for a command without a parent', () => {
		const usageString = UsageString({
			current: current as ICliNode['current'],
			parents: [],
		});
		expect(usageString).toMatch(messageArgGroup.description!);
		expect(usageString).toMatchSnapshot();
	});

	it('Creates a usage string for a command without a parent branch', () => {
		const usageString = UsageString({
			current: current as ICliNode['current'],
			parents: [branch],
		});
		expect(usageString).toMatchSnapshot();
	});

	it('Does not write usage for named argGroups if there are none', () => {
		const fooCommand = CliCommand({ name: 'foo', action() {} });
		const usageString = UsageString({ current: fooCommand, parents: [] });
		expect(usageString).toMatchSnapshot();
	});
});
