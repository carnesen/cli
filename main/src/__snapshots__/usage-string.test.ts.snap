// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsageString Creates a usage string for a branch 1`] = `
"Usage: cli <subcommand> ...

   This is a CLI

Subcommands:

   echo"
`;

exports[`UsageString Creates a usage string for a command without a parent 1`] = `
"Usage: echo [<word>] [<options>] -- <str>

<word>:

   A word

Options:

   [--message <str>] : A string message please

<str>:

   Another word"
`;

exports[`UsageString Creates a usage string for a command without a parent branch 1`] = `
"Usage: cli echo [<word>] [<options>] -- <str>

<word>:

   A word

Options:

   [--message <str>] : A string message please

<str>:

   Another word"
`;

exports[`UsageString Does not write usage for named parsers if there are none 1`] = `"Usage: foo"`;
